Model/Invoice/StatusUpdater.php

```diff
@@ -0,0 +1,142 @@
+     * @var InvoiceSender
+     */
+    protected $invoiceSender;
+
+    /**
+     * @var InvoiceRepositoryInterface
+     */
+    protected $invoiceRepository;
+
+    /**
+     * StatusUpdater constructor.
+     * @param OrderRepositoryInterface $orderRepository
+     * @param OrderStatusHistoryInterfaceFactory $orderStatusHistoryInterfaceFactory
+     * @param Config $configHelper
+     * @param InvoiceService $invoiceService
+     * @param Transaction $transaction
+     * @param InvoiceSender $invoiceSender
+     * @param InvoiceRepositoryInterface $invoiceRepository
+     */
+    public function __construct(
+        OrderRepositoryInterface $orderRepository,
+        OrderStatusHistoryInterfaceFactory $orderStatusHistoryInterfaceFactory,
+        Config $configHelper,
+        InvoiceService $invoiceService,
+        Transaction $transaction,
+        InvoiceSender $invoiceSender,
+        InvoiceRepositoryInterface $invoiceRepository
+    ) {
+        $this->orderRepository = $orderRepository;
+        $this->orderStatusHistoryInterfaceFactory = $orderStatusHistoryInterfaceFactory;
+        $this->configHelper = $configHelper;
+        $this->invoiceService = $invoiceService;
+        $this->transaction = $transaction;
+        $this->invoiceSender = $invoiceSender;
+        $this->invoiceRepository = $invoiceRepository;
+    }
+
+    /**
+     * Determinate the order is virtual and the auto invoice is setting is enabled
+     * @param Order $order
+     * @return bool
+     */
+    private function canInvoice(Order $order): bool
+    {
+        return (
+            $order->getIsVirtual()
+            && $order->canInvoice()
+            && $this->configHelper->getValue('can_auto_invoice_virtual')
+        );
+    }
+
+    /**
+     * @param Order $order
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function createInvoice(Order $order)
+    {
+        if ($this->canInvoice($order)) {
+            $invoice = $this->invoiceService->prepareInvoice($order);
+            $invoice->register();
+            $invoice->setState(Order\Invoice::STATE_PAID);
+            $this->invoiceRepository->save($invoice);
+            $transactionSave = $this->transaction->addObject(
+                $invoice
+            )->addObject(
+                $invoice->getOrder()
+            );
+            $transactionSave->save();
+            $this->invoiceSender->send($invoice);
+
+            /** @var OrderRepositoryInterface|History $history */
+            $history = $this->orderStatusHistoryInterfaceFactory->create();
+            $history->setComment(__('Notified customer about invoice #%1.', $invoice->getId()));
+            $history->setIsVisibleOnFront(0);
+            $history->setIsCustomerNotified(1);
+            $history->setStatus(Order::STATE_COMPLETE);
+
+            $order->addStatusHistory($history);
+            $order->setStatus(Order::STATE_COMPLETE);
+            $this->orderRepository->save($order);
```

Comment:
At this point this is a second order saving in the success flow. This method createInvoice has called at the end of the Order's StatusUpdater::success where the order already has been saved.

Is there any reason why is get two saves? If order save is required after invoice saving then I think this createInvoice could be easily before the order saving.

Reply:
That is a good point. shall I pass the order as a pointer?

Reply:
I think it's not necessary, the $order already points back because objects are reference types.